File: app/src/main/java/com/coyni/pos/app/baseclass/BaseActivity.java
package com.coyni.pos.app.baseclass;

import android.app.Dialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.ContextThemeWrapper;
import android.view.Gravity;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.coyni.pos.app.R;
import com.coyni.pos.app.model.appupdate.AppUpdateResp;
import com.coyni.pos.app.utils.LogUtils;
import com.coyni.pos.app.utils.MyApplication;
import com.coyni.pos.app.utils.Utils;
import com.vt.kotlinexamples.retrofit_network.viewmodel.CommonViewModel;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okio.ByteString;

public abstract class BaseActivity extends AppCompatActivity {

    public final String TAG = getClass().getName();
    private Dialog dialog;
    private MyApplication myApplication;
    CommonViewModel commonViewModel;

    public Boolean isBaseBiometric = false, isAccess = false, isMerchantHide = false;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            LogUtils.Companion.d(TAG, getClass().getName());
            myApplication = (MyApplication) getApplicationContext();
            commonViewModel = CommonViewModel.Companion.getInstance(this);

//            runOnUiThread(() -> {
            commonViewModel.getAppUpdateRespMutableLiveData().observe(this, new Observer<AppUpdateResp>() {
                @Override
                public void onChanged(AppUpdateResp appUpdateResp) {
                    try {
                        if (appUpdateResp == null) {
                            return;
                        }
                        if (appUpdateResp.getData() != null) {
                            String version = getPackageManager().getPackageInfo(BaseActivity.this.getPackageName(), 0).versionName;
                            int versionCode = getPackageManager().getPackageInfo(getPackageName(), 0).versionCode;
                            int versionName = Integer.parseInt(version.replace(".", ""));

                        }
//                        else if (appUpdateResp.getError() != null && appUpdateResp.getError().getErrorCode().equals(getString(R.string.accessrestrictederrorcode))) {
//                            showAccessRestricted();
//                        }
                    } catch (Exception ex) {
                        ex.printStackTrace();
                    }
                }
            });
//            });

        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    public void showProgressDialog() {
        showProgressDialog("Loading");
    }

    public void showProgressDialog(String message) {
        showProgressDialog(message, true, this);
    }

    public void showProgressDialog(String message, boolean isCanceledOnTouchOutside, BaseActivity baseActivity) {
        if (dialog != null && dialog.isShowing()) {
            return;
        }

        dialog = new Dialog(baseActivity);
        dialog.getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        dialog.setContentView(R.layout.loader);
        dialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);

        TextView loaderMsg = dialog.findViewById(R.id.loaderText);
        loaderMsg.setText(message);

        Window window = dialog.getWindow();
        window.setLayout(WindowManager.LayoutParams.WRAP_CONTENT, WindowManager.LayoutParams.WRAP_CONTENT);

        WindowManager.LayoutParams wlp = window.getAttributes();

        wlp.gravity = Gravity.CENTER;
        wlp.flags &= WindowManager.LayoutParams.FLAG_DIM_BEHIND;
        window.setAttributes(wlp);

        dialog.setCanceledOnTouchOutside(isCanceledOnTouchOutside);
        dialog.show();
    }

    public void dismissDialog() {
        if (dialog != null) {
            dialog.dismiss();
        }
    }


}

File: app/src/main/java/com/coyni/pos/app/baseclass/BaseActivity.kt
package com.coyni.pos.app.baseclass

import android.app.Dialog
import android.content.Intent
import android.graphics.Rect
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.view.Window
import android.view.WindowManager
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.coyni.pos.app.R
import com.coyni.pos.app.utils.LogUtils.Companion.d
import com.coyni.pos.app.utils.Utils
import com.coyni.pos.app.viewmodel.CommonViewModel

abstract class BaseActivity : AppCompatActivity() {
    val TAG = javaClass.name
    var dialog: Dialog? = null
    var commonViewModel: CommonViewModel? = null
    var isBaseBiometric = false
    var isAccess = false
    var isMerchantHide = false
    var mLastClickTime = 0L
    var isKeyboardVisible: Boolean = false
    lateinit var decorView: View
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        try {
            d(TAG, javaClass.name)
            commonViewModel = ViewModelProvider(this).get(CommonViewModel::class.java)

//            runOnUiThread(() -> {
            commonViewModel!!.appUpdateRespMutableLiveData.observe(this, Observer { appUpdateResp ->
                try {
                    if (appUpdateResp == null) {
                        return@Observer
                    }
                    if (appUpdateResp.data != null) {
                        val version = packageManager.getPackageInfo(
                            this@BaseActivity.packageName, 0
                        ).versionName
                        val versionCode = packageManager.getPackageInfo(packageName, 0).versionCode
                        val versionName = version.replace(".", "").toInt()
                    }
                    //                        else if (appUpdateResp.getError() != null && appUpdateResp.getError().getErrorCode().equals(getString(R.string.accessrestrictederrorcode))) {
                    //                            showAccessRestricted();
                    //                        }
                } catch (ex: Exception) {
                    ex.printStackTrace()
                }
            })
            //            });
        } catch (e: Exception) {
            e.printStackTrace()
        }

        initKeyboardListener()
    }

    open fun fragmentNavigation(action: String?, value: String?) {}

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
    }

    @JvmOverloads
    fun showProgressDialog(
        message: String? = "Loading",
        isCanceledOnTouchOutside: Boolean = true,
        baseActivity: BaseActivity? = this
    ) {
        if (dialog != null && dialog!!.isShowing) {
            return
        }
        dialog = Dialog(baseActivity!!)
        dialog!!.window!!.requestFeature(Window.FEATURE_NO_TITLE)
        dialog!!.setContentView(R.layout.loader)
        dialog!!.window!!.setBackgroundDrawableResource(android.R.color.transparent)
        val loaderMsg = dialog!!.findViewById<TextView>(R.id.loaderText)
        loaderMsg.text = message
        val window = dialog!!.window
        window!!.setLayout(
            WindowManager.LayoutParams.WRAP_CONTENT, WindowManager.LayoutParams.WRAP_CONTENT
        )
        val wlp = window.attributes
        wlp.gravity = Gravity.CENTER
        wlp.flags = wlp.flags and WindowManager.LayoutParams.FLAG_DIM_BEHIND
        window.attributes = wlp
        dialog!!.setCanceledOnTouchOutside(isCanceledOnTouchOutside)
        dialog!!.show()
    }

    fun dismissDialog() {
        if (dialog != null) {
            dialog!!.dismiss()
        }
    }

    private fun initKeyboardListener() {
        // Get the current window view
        decorView = window.decorView
        // Set a ViewTreeObserver on the window view
        decorView.getViewTreeObserver().addOnGlobalLayoutListener(OnGlobalLayoutListener {

            // Get the current visible display frame of the window
            val rect = Rect()
            decorView.getWindowVisibleDisplayFrame(rect)

            // Calculate the screen height and keyboard height
            val screenHeight: Int = decorView.getRootView().getHeight()
            val keyboardHeight = screenHeight - rect.bottom

            // Check if the keyboard is visible
            isKeyboardVisible = keyboardHeight > screenHeight * 0.15
            Utils.isKeyboardVisible = isKeyboardVisible
        })
    }

    open fun getKeyboardVisible(): Boolean? {
        return isKeyboardVisible
    }

}